pisces/data_sets.py

```diff
@@ -327,6 +334,440 @@ def find_data_sets(cls, root: str | Path) -> Dict[str, 'DataSetObject']:
+            acc_next_time = accel[acc_next_index, 0]
+            acc_next_index += 1
+        
+        # 1. check for lots of missing time
+        # 2. check for very low sampling rate
+        if ((acc_next_time - acc_last_time) < min_epoch_covered) \
+            or (acc_next_index - acc_last_index < min_samples_per):
+            psg_gap_indices.append(psg_index)
+        
+        # set up for next iteration
+        acc_last_time = acc_next_time
+        acc_last_index = acc_next_index
     
+    psg[np.array(psg_gap_indices), 1] = -1
 
+    return psg
 
+
+def apply_gausian_filter(df: pl.DataFrame, sigma: float = 1.0, overwrite: bool = False) -> pl.DataFrame:
+    data_columns = df.columns[1:]  # Adjust this to match your data column indices
+    # Apply Gaussian smoothing to each data column
+    for col in data_columns:
+        new_col_name = f"{col}_smoothed" if not overwrite else col
+        df = df.with_columns(
+            pl.Series(gaussian_filter1d(df[col].to_numpy(), sigma)).alias(new_col_name)
+        )
+    return df
+
+
+def fill_gaps_in_accelerometer_data(acc: pl.DataFrame, smooth: bool = False, final_sampling_rate_hz: int | None = None) -> np.ndarray:
+    # median sampling rate (to account for missing data)
+    sampling_period_s = acc[acc.columns[0]].diff().median() # 1 / sampling_rate_hz
+    
+    # Step 0: Save the original 'timestamp' column as 'timestamp_raw'
+    acc_resampled = acc.with_columns(acc[acc.columns[0]].alias('timestamp'))
+
+    #TODO: Check non int sampling rates
+    # if isinstance(final_sampling_rate_hz, int):
+    final_rate_sec = 1 / final_sampling_rate_hz
+    print(f"resampling to {final_sampling_rate_hz}Hz ({final_rate_sec:0.5f}s) from {int(1/sampling_period_s)} Hz ({sampling_period_s:0.5f}s)")
+    # make a new data frame with the new timestamps
+    # do this using linear interpolation
+
+    median_time = acc_resampled['timestamp'].to_numpy()
+    final_timestamps = np.arange(median_time.min(), median_time.max() + final_rate_sec, final_rate_sec)
+    median_data = acc_resampled[:, 1:4].to_numpy()
+    new_data = np.zeros((final_timestamps.shape[0], median_data.shape[1]))
+    for i in range(median_data.shape[1]):
+        new_data[:, i] = np.interp(final_timestamps, median_time, median_data[:, i])
+    acc_resampled = pl.DataFrame({
+        'timestamp': final_timestamps, 
+        **{
+            acc_resampled.columns[i+1]: new_data[:, i] 
+            for i in range(new_data.shape[1])
+        }})
+
+    if smooth:
+        acc_resampled = apply_gausian_filter(acc_resampled, overwrite=True)
+
+    return acc_resampled
+
+# %% ../nbs/01_data_sets.ipynb 18
+class DataProcessor:
+    def __init__(self, 
+                 data_set: DataSetObject,
+                 model_input: ModelInput,
+                 output_feature: str='psg',
+                 output_type: ModelOutputType=ModelOutputType.WAKE_LIGHT_DEEP_REM,
+                 psg_type: PSGType=PSGType.HAS_N4,
```

Comment:
I wonder if `.NO_N4` is a better default value because it has been discontinued in the literature.

Reply:
Yep, you're right